src/db.c

```diff
@@ -916,11 +983,20 @@ void scanGenericCommand(client *c, robj *o, unsigned long cursor) {
             i += 2;
         } else if (!strcasecmp(c->argv[i]->ptr, "type") && o == NULL && j >= 2) {
             /* SCAN for a particular type only applies to the db dict */
-            typename = c->argv[i+1]->ptr;
+            sds typename = c->argv[i + 1]->ptr;
+            if (sdslen(typename) == 0) {
+                addReplyErrorObject(c, shared.syntaxerr);
+                return;
+            }
```

Comment:
any reason to handle this error separately from the one below?
i see this one isn't tested or documented in the top comment, but can't we just drop it?

Reply:
is that for the expireIfNeeded call?
if so, how about this:
```diff
--- a/src/db.c
+++ b/src/db.c
@@ -1705,6 +1705,7 @@ int keyIsExpired(redisDb *db, robj *key) {
  * The return value of the function is 0 if the key is still valid,
  * otherwise the function returns 1 if the key is expired. */
 int expireIfNeeded(redisDb *db, robj *key, int flags) {
+    int static_key = key->refcount == OBJ_STATIC_REFCOUNT;
     if (server.lazy_expire_disabled) return 0;
     if (!keyIsExpired(db,key)) return 0;
 
@@ -1736,8 +1737,12 @@ int expireIfNeeded(redisDb *db, robj *key, int flags) {
      * will have failed over and the new primary will send us the expire. */
     if (isPausedActionsWithUpdate(PAUSE_ACTION_EXPIRE)) return 1;
 
+    /* create a heap robj if we only have static one */
+    if (static_key) key = createStringObject(key->ptr, sdslen(key->ptr));
+
     /* Delete the key */
     deleteExpiredKeyAndPropagate(db,key);
+    if (static_key) decrRefCount(key);
     return 1;
 }
```

Reply:
So, it means that if the key have been expired, we need copy the key string twice. But considering that the percentage of stale keys is usually  small, I think this is a good compromise. Thanks for your wonderful comment!

Reply:
seems that i commented on the wrong thread.
this was meant for https://github.com/redis/redis/pull/12209#discussion_r1233370729
i.e. the reason we create a string object was for expiry test, so we have a better solution.

while on that subject, maybe we can even avoid sdsdup?
it would be harder for the other key types (HSCAN, SSCAN, ZSCAN), but at least for normal keyspace SCAN we can easily avoid copying the string.
it'll be a little bit ugly, but i think the performance boost is worth it, if we add a few comments to clear that up.

anyway, back to the original purpose of this thread:
> any reason to handle this error separately from the one below?
i see this one isn't tested or documented in the top comment, but can't we just drop it?

Reply:
sounds good. then we'll just get the plain `unknown type name` error.